### PR TITLE
Immersive Portals!

### PR DESCRIPTION
**Pull request by Jeryn99**
This pull request was worked on by myself and @thejudge156 

It supplies Immersive Portals support for the mod for both the shell and the interior.
Shells that do not have any portal information given, will operate without portals like the mods default behaviour.

Fabric users simply need to install any version of Immersive Portals and Fabric API, but Forge users will need to use Immersive Portals 2.3.2 and onwards; using earlier versions will cause the game to close to avoid users pointing out a crash where adding a dimension via code would crash the game with a crash report.

The compatibility will be optional via the implemented config file (the only exception being on Fabric, where dimensions NEED to be handled via Immersive Portals)

### CODE REVIEW

**Review by CommandrMoose — COMMENTED** [review]
Feels way more solid than it has before. I have found a few bugs that are either in relation to or are more present when using this feature.

BUG: Leaving the game and rejoining whilst inside the cave makes the growth door no longer function.
BUG: Placing a door inside the root cave creates a portal for the growth shell
BUG: Default shell's portal is not aligned with the interior view - doors are distinctively lower inside when looking from the outside
BUG: You cannot view the external shell's model whilst inside - doors that open outwards are not visible.
BUG: Door sounds can be heard twice when the door is closed
BUG: Throwing items through the door will cause them to spaz out
BUG: Breaking the shell in creative causes the model to no longer render until the next interaction.

Before merging, please make sure that everything is operational and works as intended by building and using the TARDIS without immersive portals installed.

File: common/src/main/java/whocraft/tardis_refined/compat/portals/ImmersivePortals.java [hunk 1]
```diff
@@ -0,0 +1,259 @@
...
+        TardisEvents.DOOR_OPENED_EVENT.register(ImmersivePortals::createPortals);
+        TardisEvents.DOOR_CLOSED_EVENT.register(ImmersivePortals::destroyPortals);
+        TardisEvents.SHELL_CHANGE_EVENT.register((operator, theme) -> {
+            ImmersivePortals.destroyPortals(operator);
+            if (operator.getInternalDoor().isOpen()) {
+                ImmersivePortals.createPortals(operator);
+            }
+        });
+
+        themeToOffsetMap.put(ShellTheme.FACTORY, List.of(new Vec3(0.499, 0, 0),
```
[the remark this answers is not in the document]
Comment (Suff99) [reply common/src/main/java/whocraft/tardis_refined/compat/portals/ImmersivePortals.java:72]: Done

File: common/src/main/java/whocraft/tardis_refined/compat/portals/TARDISPortalData.java [hunk 2]
```diff
@@ -0,0 +1,7 @@
+package whocraft.tardis_refined.compat.portals;
+
+import java.util.UUID;
+
+public interface TARDISPortalData {
```
[the remark this answers is not in the document]
Comment (Suff99) [reply common/src/main/java/whocraft/tardis_refined/compat/portals/TARDISPortalData.java:5]: Renamed

File: forge/src/main/java/whocraft/tardis_refined/common/capability/forge/TardisLevelOperatorImpl.java [hunk 3]
```diff
@@ -59,6 +59,12 @@ public static void onLevelTick(TickEvent.LevelTickEvent event) {
     }
 
     public static Optional<TardisLevelOperator> get(ServerLevel level) {
+
+        if(level == null) {
+            System.out.println("what the fuck");
```
[the remark this answers is not in the document]
Comment (Suff99) [reply forge/src/main/java/whocraft/tardis_refined/common/capability/forge/TardisLevelOperatorImpl.java:64]: Removed

File: common/src/main/java/whocraft/tardis_refined/client/model/blockentity/shell/ShellModelCollection.java [hunk 4]
```diff
@@ -2,6 +2,7 @@
 
 import net.minecraft.client.Minecraft;
 import whocraft.tardis_refined.client.ModelRegistry;
+import whocraft.tardis_refined.client.model.blockentity.console.ConsoleModelCollection;
```
[the remark this answers is not in the document]
Comment (Suff99) [reply common/src/main/java/whocraft/tardis_refined/client/model/blockentity/shell/ShellModelCollection.java:5]: Removed

File: common/src/main/java/whocraft/tardis_refined/common/blockentity/door/GlobalDoorBlockEntity.java [hunk 5]
```diff
@@ -32,7 +33,7 @@ public void onRightClick(BlockState blockState, TardisInternalDoor door, Player
                     cap.setDoorClosed(true);
                     return;
                 }
-                if (!cap.getControlManager().isInFlight() && !door.locked()) {
+                if (!cap.getControlManager().isInFlight() && !door.locked() && serverLevel.dimensionTypeId().equals(DimensionTypes.TARDIS)) {
```
[the remark this answers is not in the document]
Comment (Suff99) [reply common/src/main/java/whocraft/tardis_refined/common/blockentity/door/GlobalDoorBlockEntity.java:36]: Removed

File: common/src/main/java/whocraft/tardis_refined/client/model/blockentity/door/VendingMachineDoorModel.java [hunk 6]
```diff
@@ -50,7 +50,7 @@ public ModelPart root() {
 
     @Override
     public void setDoorPosition(boolean open) {
-        this.door.xRot = (open) ? 1.6f : 0;
+        this.door.xRot = (open) ? -1.6f : 0;
```
[the remark this answers is not in the document]
Comment (Suff99) [reply common/src/main/java/whocraft/tardis_refined/client/model/blockentity/door/VendingMachineDoorModel.java:53]: Fixed

File: common/src/main/java/whocraft/tardis_refined/client/model/blockentity/door/PresentDoorModel.java [hunk 7]
```diff
@@ -66,7 +66,7 @@ public ModelPart root() {
 
 	@Override
 	public void setDoorPosition(boolean open) {
-		this.door.xRot = (open) ? 1.5f:0;
+		this.door.xRot = (open) ? -1.5f:0;
```
[the remark this answers is not in the document]
Comment (Suff99) [reply common/src/main/java/whocraft/tardis_refined/client/model/blockentity/door/PresentDoorModel.java:69]: Fixed

File: common/src/main/java/whocraft/tardis_refined/client/model/blockentity/door/PhoneBoothDoorModel.java [hunk 8]
```diff
@@ -77,7 +77,7 @@ public void setupAnim(Entity entity, float f, float g, float h, float i, float j
 
 	@Override
 	public void setDoorPosition(boolean open) {
-		this.Door2.yRot = (open) ? -1.75f : 0;
+		this.Door2.yRot = (open) ? 1.75f : 0;
```
[the remark this answers is not in the document]
Comment (Suff99) [reply common/src/main/java/whocraft/tardis_refined/client/model/blockentity/door/PhoneBoothDoorModel.java:80]: Fixed

**Comment by Suff99** [comment]
- [ ] BUG: Leaving the game and rejoining whilst inside the cave makes the growth door no longer function.
- [ ] BUG: Placing a door inside the root cave creates a portal for the growth shell
- [ ] BUG: Default shell's portal is not aligned with the interior view - doors are distinctively lower inside when looking from the outside
- [ ] BUG: You cannot view the external shell's model whilst inside - doors that open outwards are not visible.
- [ ] BUG: Door sounds can be heard twice when the door is closed
- [ ] BUG: Throwing items through the door will cause them to spaz out
- [ ] BUG: Breaking the shell in creative causes the model to no longer render until the next interaction.

File: common/build.gradle [hunk 9]
```diff
@@ -5,6 +5,17 @@ dependencies {
...
+
+    modCompileOnlyApi('com.github.iPortalTeam.ImmersivePortalsMod:imm_ptl_core:v2.3.1-1.19') {
+        exclude(group: "net.fabricmc.fabric-api")
+        transitive(false)
+    }
+    modCompileOnlyApi('com.github.iPortalTeam.ImmersivePortalsMod:q_misc_util:v2.3.1-1.19') {
+        exclude(group: "net.fabricmc.fabric-api")
+        transitive(false)
+    }
+
```
Comment (50ap5ud5) [inline common/build.gradle:18]: Would be good to set a gradle variable for the IP dependencies

File: common/src/main/java/whocraft/tardis_refined/common/tardis/manager/TardisInteriorManager.java [hunk 10]
```diff
@@ -89,21 +105,19 @@ public CompoundTag saveData(CompoundTag tag) {
...
-            tag.putString(NbtConstants.TARDIS_IM_PREPARED_THEME, "");
-        }
-
+        tag.putString(NbtConstants.TARDIS_IM_PREPARED_THEME, this.preparedTheme != null ? this.preparedTheme.id : "");
+        tag.putString(NbtConstants.TARDIS_IM_CURRENT_THEME, this.currentTheme.id);
         return tag;
     }
 
     public void loadData(CompoundTag tag) {
+        System.out.println(tag);
```
Comment (50ap5ud5) [inline common/src/main/java/whocraft/tardis_refined/common/tardis/manager/TardisInteriorManager.java:114]: Remove prints

File: fabric/src/main/java/whocraft/tardis_refined/fabric/events/ModEvents.java [hunk 11]
```diff
@@ -38,12 +56,43 @@ public static void addCommonEvents() {
...
+        JsonArray jsonArray = new JsonParser().parse(json).getAsJsonArray();
+        for (int i = 0; i < jsonArray.size(); i++) {
+            JsonObject jsonObject = jsonArray.get(i).getAsJsonObject();
+            String uuid = jsonObject.get("uuid").getAsString();
+            uuids.add(uuid);
+        }
+        return uuids;
+    }
+
+    public static JsonElement getResponse(URL url) throws IOException {
```
Comment (50ap5ud5) [inline fabric/src/main/java/whocraft/tardis_refined/fabric/events/ModEvents.java:84]: What is this used for? Seems to not be called anywhere.

**Review by 50ap5ud5 — CHANGES_REQUESTED** [review]
Amazing work!

**Review by 50ap5ud5 — APPROVED** [review]

**Review by CommandrMoose — APPROVED** [review]

**Comment by Noklle** [comment]
Looks stunning, but I came across a weird issue

BUG: The collision of the door block prevents entry and exit. This can be circumvented in the overworld by jumping on top of the block, upon which you fall into it and can walk into the TARDIS, but this doesn't work when trying to exit the TARDIS. Works fine in spectator mode, given the lack of collision.

**Comment by Jeryn99** [comment]
It's been fixed pending release

**Pull request merged**